### ISSUE TITLE
About random factor in the embedding/tokenization process

### ISSUE THREAD

**Issue by josecar24** [MAGICS-LAB/DNABERT_2#91]
I'm wondering if other users are also experiencing this situation: when I try to convert the same DNA sequence into a tensor multiple times using DNABERT2, the hidden state converted several times in a row comes out with different values. Is this because there is a random factor in the conversion process? I wonder how I can find it and make sure that the tensor I convert from the same DNA sequence no longer varies randomly?
For instance, from the same DNA sequence "GGCTGCGCGTACATGGGTCGT", I had ran DNABERT2 twice and the output tensors below:
<img width="451" alt="Tensor_1" src="https://github.com/MAGICS-LAB/DNABERT_2/assets/112451709/179af4ba-44b2-4cba-b489-5cfb4d490dc5">
<img width="601" alt="Tensor_2" src="https://github.com/MAGICS-LAB/DNABERT_2/assets/112451709/ec6ac15b-9048-479f-9141-f8694e8fbe98">

**Comment by Zhihan1996** [comment]
Can you share the script you used?

**Comment by josecar24** [comment]
Sure, the original notebook could not be uploaded. But the script I used is the original one `dna = "GGCTGCGCGTACATGGGTCGT"
inputs = tokenizer(dna, return_tensors = 'pt')["input_ids"]
hidden_states = model(inputs)[0] # [1, sequence_length, 768]

# embedding with mean pooling
embedding_mean = torch.mean(hidden_states[0], dim=0)
print(embedding_mean.shape) 

# embedding with max pooling
embedding_max = torch.max(hidden_states[0], dim=0)[0]
print(embedding_max.shape) 

print(hidden_states)`

<img width="953" alt="截屏2024-06-09 下午12 46 29" src="https://github.com/MAGICS-LAB/DNABERT_2/assets/112451709/71e719f6-f160-4a1b-bbff-5f7c644f2f43">

**Comment by josecar24** [comment]
> Can you share the script you used?

Thank you for your quick reply. I'm sorry for any inconvenience caused, since I'm a cross-disciplinary student, my expression may have been a bit confusing at first.

**Comment by Zhihan1996** [comment]
Thanks for sharing the script. I am happy to help!

There could be two possible causes:

1. The model is not set to evaluation mode so the dropout may applies. Please try `model = model.eval()` before generating the hidden states.
2. I am not very sure whether the `from_config` API loads the pre-trained model weight or randomly initialize it when the config itself is from_pretrained. You may try to load the model with

```
! pip install transformers=4.28

from transformers import AutoTokenizer, AutoModel

tokenizer = AutoTokenizer.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
model = AutoModel.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
```

**Comment by josecar24** [comment]
> Thanks for sharing the script. I am happy to help!
> 
> There could be two possible causes:
> 
> 1. The model is not set to evaluation mode so the dropout may applies. Please try `model = model.eval()` before generating the hidden states.
> 2. I am not very sure whether the `from_config` API loads the pre-trained model weight or randomly initialize it when the config itself is from_pretrained. You may try to load the model with
> 
> ```
> ! pip install transformers=4.28
> 
> from transformers import AutoTokenizer, AutoModel
> 
> tokenizer = AutoTokenizer.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
> model = AutoModel.from_pretrained("zhihan1996/DNABERT-2-117M", trust_remote_code=True)
> ```

Thank you for your kind assistance. 
1. Adding a model = model.eval() to the original code where you pointed out does help!
2. I am afraid loading the model this way will encounter this issue (https://huggingface.co/zhihan1996/DNABERT-2-117M/discussions/26). Actually, the original code for model load works well 2 month ago. But the loading issue happens when I saw that discussion, so I changed the code for model loading following that discussion.
![Uploading 截屏2024-06-09 下午1.04.12.png…]()

**Comment by Zhihan1996** [comment]
Great, glad to know it helps. 

As for `Actually, the original code for model load works well 2 month ago. But the loading issue happens when I saw that discussio`, It may result from some mistakes I made that time. Thanks for letting me know. I will look into it later.

**Comment by josecar24** [comment]
Hello Dr. ZHOU(@Zhihan1996), 

For the random problem, I found the problem still exists for same DNA sequence if I reload the model by running the first code cell. For instance, If I run the second code cell twice in a row directly after the model is loaded. The output tensor will be consistent due to 'model = model.eval()' as you suggested. But if I run the first code cell again between running the second and third code cells, and the random problem will reappear (and not be solved with my addition of model.eval after model loaded in code cell one).
Is that designed to be different outcome for operations between batches? Is there any possible solution to keep the output tensor from the same DNA sequence always the same please? Appreciate again!


<img width="812" alt="截屏2024-06-17 下午12 51 41" src="https://github.com/MAGICS-LAB/DNABERT_2/assets/112451709/a5a7d9f3-0045-4f4d-aeaf-b585ba23d0c2">
